src/dlgPackageExporter.cpp

```diff
@@ -268,64 +317,230 @@ void dlgPackageExporter::slot_export_package()
+                if (zip_dir_add(archive, directoryName.toStdString().c_str(), ZIP_FL_ENC_UTF_8) == -1) {
+                    displayResultMessage(tr("Failed to add directory \"%1\" to module (archive) file, error is: \"%2\".")
+                                         .arg(directoryName, QString::fromUtf8(zip_strerror(archive))), false);
+                    zip_close(archive);
+                    isOk = false;
+                }
+            }
+
+            // Process the config and the file containing the Mudlet triggers,
+            // etc. specially so they are inserted first and last respectively:
+            if (isOk) {
+                // Apparently it is permissable for there NOT to be a config.lua
+                // file in the manufactured module (i.e. the user is allowed to
+                // manually remove it even though we now request they do not).
+                // If we enhance modules in the future to store more data in
+                // that file I think this will no longer be permissable and the
+                // two commented out bits of code can be restored
+                if (fileEntries.contains(QStringLiteral("config.lua"))) {
+                    if (!writeFileToZip(QStringLiteral("config.lua"), fileEntries.value(QStringLiteral("config.lua")), archive)) {
+                        /* isOk = false; */
+                    } else {
+                        fileEntries.remove(QStringLiteral("config.lua"));
+                    }
+                }/* else {
+                    displayResultMessage(tr("Required \"config.lua\" file not found to include in module (archive) file, did you remove or rename it?"), false);
+                    zip_close(archive);
+                    isOk = false;
+                }*/
+            }
+
+            QString xmlFileName = mPackageName;
+            xmlFileName.append(QLatin1String(".xml"));
+            if (isOk) {
+                QMapIterator<QString, QString> itFileName(fileEntries);
+                while (itFileName.hasNext() && isOk) {
+                    itFileName.next();
+                    if (itFileName.key() == xmlFileName) {
+                        continue;
+                    }
+
+                    if (!writeFileToZip(itFileName.key(), itFileName.value(), archive)) {
+                        zip_close(archive);
+                        isOk = false;
+                        break;
+                    }
+                }
+            }
+
+            if (isOk) {
+                if (fileEntries.contains(xmlFileName) && fileEntries.value(xmlFileName) == mXmlPathFileName) {
+                    if (!writeFileToZip(xmlFileName, mXmlPathFileName, archive)) {
+                        isOk = false;
+                    }
+
+                    // If successful will get to HERE...
+
+                } else {
+                    displayResultMessage(tr("Required \"%1\" file that contains the mudlet items chosen for the "
```

Comment:
`Required "name" file` vs. `Required file "name"`. Also run-on sentence could be split easily in 3-4 parts to ease translatability. Again not sure why we keep mentioning "module" and "module (archive)" here, as there should be a single consistent word used and for me that is package.

Reply:
I must confess that I have a conscious bias in calling `.xml` files containing a *single* type (trigger, timer, key-binding, etc.) of Mudlet item a **package** whereas the output from this code which can contain *more* than one type of Mudlet items and that can also *contain other files* a **module**.

The mentioning of `archive` is because libzip is being used to handle what is actually an `.zip` *archive* type file and in some circumstances it makes sense (to me) to say that to explain what is (or is not) happening - most people understand what a **`zip`** file is (a file that contains lots of other files that have been squashed up inside it).

Actually, `.zip` was the file extension being used for what was being produced until this PR. This one *finally* starts using the `.mpackage` extension that we have been coding for since Chris's commit https://github.com/Mudlet/Mudlet/commit/f98be22c0915bab3f26ba9e22d3834b51f9df793 in **2013**! :open_mouth:

I will take a look at the wordings though to see if I can make my sentences shorter...

Reply:
That's not correct though - the only difference between a package and a module is how it is installed in Mudlet.

Reply:
So we do not distinguish between:
+ the item that is a single XML file contain only one type of Mudler item and made from the "Export" button in the editor
+ the item that is an archive file containing at least an XML file of maybe more then one type of Mudlet item, a `config.lua` file that *currently* **\+** records the module name (but which is retained internally in the `(QString) Tree::mPackageName` and NOT the unused `(QString) Tree::mModuleName` item members) and any other files that the creator wants included and is made from the (misnamed I feel) "Package exporter (experimental)"

:question: 

**\+** - I think we can extend that file to also include other things like the creator/licence (?)/creation datetime stamp/version number/save number and - to solve https://github.com/Mudlet/Mudlet/issues/659 the helpURL link...

Reply:
That's right, we don't (and no need to!)

Reply:
But, but, but they are just **not** the same as far as I can see. :open_mouth:

For instance you can (if you were brave enough open the XML form in a text editor and hack it by hand) if you open an archive file directly in something like `notepad++.exe` you will not get anything you can use - you have to open the file with something like `7Zip.exe` and then work on the files within the archive {not sure if Windoze `explorer.exe` is clever enough to spot that an `.mpackage` file is a zip archive!}

The name that is displayed in the editor for the module (or package as you insist) {the brown folder} need not be the same as the filename it is loaded from for the case using an archive file but it is forced to be that (without the `.xml` {or `.mtrigger`}) for the simple XML case.

Reply:
All this could be discussed in wiki to some extent, but can't be extensively in a tool-tip or error message. It is only confusing most users to see "module (archive) file" and no "package" mentioned in a "package exporter" functionality...

Reply:
Reworded (to include *package* instead of *module*), broken up into more sentences, and lack of space added as a possible reason for the problem alongside that of *permissions*.

Ah:
> "... that contains the mudlet items ..."

Should I have it as **M**udlet items?